src/main/java/com/google/devtools/build/lib/rules/proto/ProtoToolchainRule.java

```diff
@@ -0,0 +1,67 @@
+// Copyright 2020 The Bazel Authors. All rights reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//    http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package com.google.devtools.build.lib.rules.proto;
+
+import static com.google.devtools.build.lib.packages.Attribute.attr;
+import static com.google.devtools.build.lib.packages.BuildType.LABEL;
+
+import com.google.common.base.Preconditions;
+import com.google.devtools.build.lib.analysis.BaseRuleClasses;
+import com.google.devtools.build.lib.analysis.RuleDefinition;
+import com.google.devtools.build.lib.analysis.RuleDefinitionEnvironment;
+import com.google.devtools.build.lib.analysis.config.ExecutionTransitionFactory;
+import com.google.devtools.build.lib.cmdline.Label;
+import com.google.devtools.build.lib.packages.Attribute;
+import com.google.devtools.build.lib.packages.RuleClass;
+
+/** Rule definition of the <code>proto_toolchain</code> rule. */
+class ProtoToolchainRule implements RuleDefinition {
+  static final String PROTO_COMPILER_ATTR_NAME = ":proto_compiler";
+
+  private final Label defaultProtoCompiler;
+
+  public ProtoToolchainRule(Label defaultProtoCompiler) {
+    this.defaultProtoCompiler = Preconditions.checkNotNull(defaultProtoCompiler);
+  }
+
+  @Override
+  public RuleClass build(RuleClass.Builder builder, final RuleDefinitionEnvironment env) {
+    return builder
+        .add(
+            attr(PROTO_COMPILER_ATTR_NAME, LABEL)
+                .cfg(ExecutionTransitionFactory.create())
+                .exec()
+                .value(
+                    Attribute.LabelLateBoundDefault.fromTargetConfiguration(
+                        ProtoConfiguration.class,
+                        defaultProtoCompiler,
+                        (rule, attributes, protoConfig) ->
+                            protoConfig.protoCompiler() != null
+                                ? protoConfig.protoCompiler()
+                                : defaultProtoCompiler)))
+        .requiresConfigurationFragments(ProtoConfiguration.class)
+        .advertiseProvider(ProtoToolchainInfo.class)
+        .build();
```

Comment:
Why don't you add Label defaultProtoCompiler as an attribute to the rule and remove it from the constructor? You can specify default value on the attribute or perhaps even better leave it blank and set it in the BUILD file.

It looks like you invested some effort to provide a design where ProtoToolchainRule can be modified by other, that is creating an abstract BazelProtoToolchainRule, however I think it is better to keep it simple without an additional abstract class. Toolchain rules usually just need to pass the data thru, and don't have much logic that would need changing. (This is different to BazelProtoLibraryRule)

Reply:
The abstraction exists so that Google can create an internal `proto_toolchain` rule that uses a different label for `protoc`. If you can achieve the same thing with Copybara transformations, I'm happy to remove the abstraction :).

Reply:
Would that be the case if the label was only specified on an attribute? I expect there to be a different BUILD file at Google internally than the new WORKSPACE suffix.

Reply:
Yes, there needs to be an internal version of what is in `tools/proto/BUILD.tools` and the new toolchain must be registered in Google's WORKSPACE file.

Note that there are currently no settable attributes on `proto_toolchain`, everything is read from the configuration fragment. Thinking on this a bit more, it's probably really is fine to just have 1 `proto_toolchain` that is identical for Bazel and Blaze. I'll update the PR.